chainerx_cc/chainerx/routines/reduction.cc

```diff
@@ -144,4 +146,49 @@ Array Cumsum(const Array& a, absl::optional<int8_t> axis) {
     return out;
 }
 
+Array Nansum(const Array& a, const OptionalAxes& axis, bool keepdims) {
+    Axes sorted_axis = internal::GetSortedAxesOrAll(axis, a.ndim());
+    Array a_masked = Where(IsNan(a), 0, a);
+    // Decide the output dtype for integral input dtype.
+    Dtype out_dtype{};
+    switch (GetKind(a_masked.dtype())) {
+        case DtypeKind::kBool:
+        case DtypeKind::kInt:  // fallthrough
+            out_dtype = Dtype::kInt64;
+            break;
+        case DtypeKind::kUInt:
+            out_dtype = Dtype::kInt64;  // TODO(niboshi): This should be kUInt64
+            break;
+        default:
+            out_dtype = a.dtype();
+    }
+
+    Array out = internal::EmptyReduced(a_masked.shape(), out_dtype, sorted_axis, keepdims, a_masked.device());
+    {
+        NoBackpropModeScope scope{};
+        a.device().backend().CallKernel<NansumKernel>(a_masked, sorted_axis, out);
+    }
+
+    BackwardBuilder bb{"nansum", a, out};
+    if (BackwardBuilder::Target bt = bb.CreateTarget(0)) {
+        bt.Define([sorted_axis, in_shape = a.shape(), keepdims](BackwardContext& bctx) {
+            const Array& gout = *bctx.output_grad();
+            const Array& input_grad = Where(IsNan(gout), 0, gout);
+            CHAINERX_ASSERT(std::is_sorted(sorted_axis.begin(), sorted_axis.end()));
+
+            if (!(in_shape.ndim() == 0 || sorted_axis.empty() || keepdims)) {
+                Shape out_shape_broadcastable = gout.shape();
+                for (auto axis : sorted_axis) {
+                    out_shape_broadcastable.insert(out_shape_broadcastable.begin() + axis, 1);
+                }
+                bctx.input_grad() = input_grad.Reshape(out_shape_broadcastable).BroadcastTo(in_shape);
+            } else {
+                bctx.input_grad() = input_grad.BroadcastTo(in_shape);
+            }
+        });
```

Comment:
```Python
if (BackwardBuilder::Target bt = bb.CreateTarget(0)) {
        bt.Define([a_tok = bb.RetainInput(0), sorted_axis, in_shape = a.shape(), keepdims](BackwardContext& bctx) {
            const Array& gout = *bctx.output_grad();
            const Array& input = bctx.GetRetainedInput(a_tok);
            Array& input_grad = bctx.input_grad();
            CHAINERX_ASSERT(std::is_sorted(sorted_axis.begin(), sorted_axis.end()));

            if (!(in_shape.ndim() == 0 || sorted_axis.empty() || keepdims)) {
                Shape out_shape_broadcastable = gout.shape();
                for (auto axis : sorted_axis) {
                    out_shape_broadcastable.insert(out_shape_broadcastable.begin() + axis, 1);
                }
                input_grad = gout.Reshape(out_shape_broadcastable).BroadcastTo(in_shape);
            } else {
                input_grad = gout.BroadcastTo(in_shape);
            }
            input_grad = Where(IsNan(input), 0, input_grad);
        });
    }
```

The backward code is wrong, `gout` never has nans, you need to retain the input and mask with it.

Reply:
@emcastillo @toslunar Thank you for the suggestions. I have fixed the backward implementation and tests. Tests now pass successfully. Please review:)